### PR TITLE
feat(Core/Misc): Implemented ObjectGuid class. Port from TrinityCore.

### LINKED ISSUE #3733: [Improvement] Guid system [$150]
Hello,
More than a year ago I was working on implementing GOMove and I noticed that we have an issue on our `.gob move` command. See here : https://github.com/azerothcore/azerothcore-wotlk/issues/1588
Fact it, `.gob move` is working as intended, but the problem is : client isn't notified of the update (new coords, rotation...).
This problem is coming from our guid system.

This is what our dear friend @Rochet2 explained to me (here is a part of our conversation) :

**Rochet2** _18/03/2019_
> that same issue was on TC before they moved to the new guid system
> .gob move just relocates the existing object and does not change the in game guid.
> however the game is coded so that a gameobject cannot be moved.
> So the only way around this is to use different guids every time a gameobject is moved.
> TC's system uses different guid in game and in DB.
> on TC the gameobject's DB guid will not change, but in game guid will
> so DB keeps having 1000, but the in game guid is something like 123, 124, 125 and so on
> However AC (and the old TC guid system) uses same guid in game and in DB. 

**Shard** _18/03/2019_
> Isn't creatures using the same guid system ? Because .npc move works

**Rochet2** _18/03/2019_
> yes, but creatures can be moved.
> gameobjects cannot
> the game client was probably coded to assume all objects are static.
> maybe its some kind of optimization

I think implementing TC's guid system will solve the issue, and some related too issues like this one (maybe?) https://github.com/azerothcore/azerothcore-wotlk/issues/2518

TC source code: https://github.com/TrinityCore/TrinityCore/tree/3.3.5

##### CURRENT BEHAVIOUR:
Check : https://github.com/azerothcore/azerothcore-wotlk/issues/1588

##### EXPECTED BLIZZLIKE BEHAVIOUR:
Player should be notified about his environment changes

##### STEPS TO REPRODUCE THE PROBLEM:
Check https://github.com/azerothcore/azerothcore-wotlk/issues/1588

#### EXTRA INFO!
Here is Rochet's commit to adapt GOMove to the new TC guid system + some info he sent to me :)

> here is the commit that removes the hack
> https://github.com/Rochet2/TrinityCore/commit/1ad9187474649b1e28a33de1013e4f5fdf01c26d#diff-d8682ec190918750ac5c6535e5ab55f1
> the way it worked was that when you move an object a new object is spawned with a guid that has a custom highguid value.
> I guess the core  could have tracked the custom guids, but the way I implemented it was that the server sends the fake guids to the client as hex. That way you can in client see what object is fake and what is real.
> The fake objects were not saved. You had to click a save button to save them, which respawned them with a real guid.
> the fake guid thing is in src/server/game/Entities/Object/ObjectGuid.h

##### AC HASH/COMMIT:
Commit : Latest :kappa:
               da304a76c9c74a4dcf0a38c88a860dc5f97f3c18

##### OPERATING SYSTEM:
Win10/Debian9

##### MODULES:
None

<bountysource-plugin>

---
There is a **[$150 open bounty](https://www.bountysource.com/issues/94495403-improvement-guid-system?utm_campaign=plugin&utm_content=tracker%2F40032087&utm_medium=issues&utm_source=github)** on this issue. Add to the bounty at [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F40032087&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### PR DESCRIPTION
**Pull request by UltraNix**
## Issues Addressed:
- Closes #3733.

## How to Test the Changes:

This is a very complex PR, so there is nothing specifically to be tested. Just log in, do your stuff (killing mobs, questing, exploring, PvPing, etc) and see if any bugs related to the PR will show up.
Especially, you should keep an eye on loading from/saving to db many things (pets, auras, cooldowns, auctions, mails, mob respawns, etc) and observe errors that can appear in logs/console.

**Connected PRs:**
https://github.com/azerothcore/mod-premium-lib/pull/4
https://github.com/azerothcore/mod-weapon-visual/pull/18
https://github.com/azerothcore/mod-npc-codebox/pull/16
https://github.com/azerothcore/mod-npc-beastmaster/pull/44
https://github.com/azerothcore/mod-morphsummon/pull/13
https://github.com/azerothcore/mod-buff-command/pull/19
https://github.com/azerothcore/mod-transmog/pull/44

**Eluna: https://github.com/ElunaLuaEngine/Eluna/pull/357**

**Is a requirement for:**
#5223

## Target Branch(es):
- [x] Master

 
## How to Test AzerothCore PRs
 
When a PR is ready to be tested, it will be marked as **[WAITING TO BE TESTED]**.

You can help by testing PRs and writing your feedback here in the PR's page on GitHub. Follow the instructions here:

http://www.azerothcore.org/wiki/How-to-test-a-PR

### CODE REVIEW

**Comment by Winfidonarleyan** [comment]
How much do you think is left?

**Comment by UltraNix** [comment]
Buildfixes only.

**Comment by Winfidonarleyan** [comment]
Okay, good luck

File: src/server/game/Entities/Player/Player.h [hunk 1]
```diff
@@ -1625,17 +1623,17 @@ class Player : public Unit, public GridObject<Player>
     [[nodiscard]] Unit* GetSelectedUnit() const;
     [[nodiscard]] Player* GetSelectedPlayer() const;
 
-    void SetTarget(uint64 /*guid*/) override { } /// Used for serverside target changes, does not apply to players
-    void SetSelection(uint64 guid);
+    void SetTarget(ObjectGuid /*guid*/ = ObjectGuid::Empty) override { } /// Used for serverside target changes, does not apply to players
```
Comment (Winfidonarleyan) [inline src/server/game/Entities/Player/Player.h:1626]: Why add `= ObjectGuid::Empty`?

Comment (UltraNix) [reply src/server/game/Entities/Player/Player.h:1626]: It's an overriden function - default is empty.

Comment (Winfidonarleyan) [reply src/server/game/Entities/Player/Player.h:1626]: Okay

File: data/sql/updates/pending_db_characters/rev_1617907126348389400.sql [hunk 2]
```diff
@@ -19,4 +19,11 @@ ALTER TABLE `auctionhouse` CHANGE `auctioneerguid` `houseid` TINYINT(3)  UNSIGNE
...
 
-ALTER TABLE `groups` CHANGE `icon1` `icon1` BIGINT UNSIGNED NOT NULL, CHANGE `icon2` `icon2` BIGINT UNSIGNED NOT NULL, CHANGE `icon3` `icon3` BIGINT UNSIGNED NOT NULL, CHANGE `icon4` `icon4` BIGINT UNSIGNED NOT NULL, CHANGE `icon5` `icon5` BIGINT UNSIGNED NOT NULL, CHANGE `icon6` `icon6` BIGINT UNSIGNED NOT NULL, CHANGE `icon7` `icon7` BIGINT UNSIGNED NOT NULL, CHANGE `icon8` `icon8` BIGINT UNSIGNED NOT NULL; 
+ALTER TABLE `groups` CHANGE `icon1` `icon1` BIGINT UNSIGNED NOT NULL;
+ALTER TABLE `groups` CHANGE `icon2` `icon2` BIGINT UNSIGNED NOT NULL;
+ALTER TABLE `groups` CHANGE `icon3` `icon3` BIGINT UNSIGNED NOT NULL;
+ALTER TABLE `groups` CHANGE `icon4` `icon4` BIGINT UNSIGNED NOT NULL;
+ALTER TABLE `groups` CHANGE `icon5` `icon5` BIGINT UNSIGNED NOT NULL;
+ALTER TABLE `groups` CHANGE `icon6` `icon6` BIGINT UNSIGNED NOT NULL;
+ALTER TABLE `groups` CHANGE `icon7` `icon7` BIGINT UNSIGNED NOT NULL;
+ALTER TABLE `groups` CHANGE `icon8` `icon8` BIGINT UNSIGNED NOT NULL;
```
Comment (FrancescoBorzi) [inline data/sql/updates/pending_db_characters/rev_1617907126348389400.sql:29]: o_O

Comment (UltraNix) [reply data/sql/updates/pending_db_characters/rev_1617907126348389400.sql:29]: Yep - this pending sql build check is kinda weird...

**Comment by UltraNix** [comment]
Ready to be reviewed/tested.

This is a very complex PR, so there is nothing specifically to be tested. Just log in, do your stuff (killing mobs, questing, exploring, PvPing, etc) and observer if any bug related to the PR will show up.
Especially, you should keep an eye on loading from/saving to db many things (pets, auras, cooldowns, auctions, mails, mob respawns, etc) and observer errors that can appear on the console/logs.

I will this PR in smaller commits for now.

**Comment by UltraNix** [comment]
Is anybody testing it?

**Comment by Winfidonarleyan** [comment]
Conflicts

**Comment by UltraNix** [comment]
Because of wrong change on master.

**Comment by FrancescoBorzi** [comment]
> Because of wrong change on master.

what wrong change?

image

**Comment by UltraNix** [comment]
Conflicts resolved.

**Comment by FrancescoBorzi** [comment]
@UltraNix unfortunately there are still some

**Comment by UltraNix** [comment]
I can't resolve the rest of them because they are the part of my changes.

**Comment by UltraNix** [comment]
But they are the part of my changes.

**Comment by UltraNix** [comment]
Conflicts resolved.

File: src/common/Dynamic/TypeContainer.h [hunk 3]
```diff
@@ -108,4 +115,34 @@ class TypeMapContainer
...
 };
+
+template<class OBJECT_TYPES, class KEY_TYPE>
+class TypeUnorderedMapContainer
+{
+public:
+    template<class SPECIFIC_TYPE>
+    bool Insert(KEY_TYPE const& handle, SPECIFIC_TYPE* obj)
+    {
+        return acore::Insert(_elements, handle, obj);
```
Comment (Winfidonarleyan) [inline src/common/Dynamic/TypeContainer.h:126]: `nullptr`

File: src/common/Dynamic/TypeContainer.h [hunk 4]
```diff
@@ -108,4 +115,34 @@ class TypeMapContainer
...
+    }
+
+    template<class SPECIFIC_TYPE>
+    bool Remove(KEY_TYPE const& handle)
+    {
+        return acore::Remove(_elements, handle, (SPECIFIC_TYPE*)NULL);
```
Comment (Winfidonarleyan) [inline src/common/Dynamic/TypeContainer.h:132]: `nullptr`

File: src/common/Dynamic/TypeContainer.h [hunk 5]
```diff
@@ -108,4 +115,34 @@ class TypeMapContainer
...
+    }
+
+    template<class SPECIFIC_TYPE>
+    SPECIFIC_TYPE* Find(KEY_TYPE const& handle)
+    {
+        return acore::Find(_elements, handle, (SPECIFIC_TYPE*)NULL);
```
Comment (Winfidonarleyan) [inline src/common/Dynamic/TypeContainer.h:138]: `nullptr`

File: src/server/game/Entities/Object/Object.cpp [hunk 6]
```diff
@@ -640,58 +610,58 @@ void Object::SetUInt64Value(uint16 index, uint64 value)
...
 
         return true;
     }
 
     return false;
 }
 
+void Object::SetGuidValue(uint16 index, ObjectGuid value)
+{
+    ASSERT(index + 1 < m_valuesCount || PrintIndexError(index, true));
```
Comment (Winfidonarleyan) [inline src/server/game/Entities/Object/Object.cpp:654]: maybe new line after this?

**Comment by Winfidonarleyan** [comment]
Ping me, after resolve conflicts

**Comment by UltraNix** [comment]
@Winfidonarleyan Done.

**Comment by FrancescoBorzi** [comment]
To all hands here, we need to have this properly tested **before** it gets to master. Please test this, preferably with players too.

cc @azerothcore/testers @azerothcore/modules @azerothcore/developers @mpfans @pklloveyou @Footman @disclosurez

**Comment by Yehonal** [comment]
NOTE: It will break the compatibility with almost ALL modules. Just keep it in mind

**Comment by FrancescoBorzi** [comment]
@Yehonal even after this ?


> Connected PRs:
azerothcore/mod-premium-lib#4
azerothcore/mod-weapon-visual#18
azerothcore/mod-npc-codebox#16
azerothcore/mod-npc-beastmaster#44
azerothcore/mod-morphsummon#13
azerothcore/mod-buff-command#19
azerothcore/mod-transmog#44

**Comment by Yehonal** [comment]
> @Yehonal even after this ?
> 
> > Connected PRs:
> > [azerothcore/mod-premium-lib#4](https://github.com/azerothcore/mod-premium-lib/pull/4)
> > [azerothcore/mod-weapon-visual#18](https://github.com/azerothcore/mod-weapon-visual/pull/18)
> > [azerothcore/mod-npc-codebox#16](https://github.com/azerothcore/mod-npc-codebox/pull/16)
> > [azerothcore/mod-npc-beastmaster#44](https://github.com/azerothcore/mod-npc-beastmaster/pull/44)
> > [azerothcore/mod-morphsummon#13](https://github.com/azerothcore/mod-morphsummon/pull/13)
> > [azerothcore/mod-buff-command#19](https://github.com/azerothcore/mod-buff-command/pull/19)
> > [azerothcore/mod-transmog#44](https://github.com/azerothcore/mod-transmog/pull/44)

There are dozens of other modules that are using the GUIDs.

**Comment by FrancescoBorzi** [comment]
ah sure, it would be nice if people can test it together with the modules they are typically using so we can prepare PRs in advance just like UltraNix already did with some modules

**Comment by Yehonal** [comment]
> ah sure, it would be nice if people can test it together with the modules they are typically using so we can prepare PRs in advance just like UltraNix already did with some modules

@UltraNix wouldn't be possible to offer a "legacy" support? by keeping the low/mid/high variables and the related method to keep the compatibility with modules?

It will allow us to slowly migrate to the new GUID system and eventually remove the legacy layer in future

**Comment by UltraNix** [comment]
@Yehonal I made PRs in every modul that uses guids or is changes by that PR. Give me an example of modul that I forgot about?

**Comment by Yehonal** [comment]
> @Yehonal I made PRs in every modul that uses guids or is changes by that PR. Give me an example of modul that I forgot about?
> And no - there is no possibility to offer "legacy" code or support, because of confilicts between modules and core. Sink or swim.

There are modules that are not published in our organization and I'm not sure about eluna as well. Have you checked that module? do the .lua files need to be updated?

If it's not possible to keep those legacy methods/variables then we need to notify everyone before merge this

**Comment by FrancescoBorzi** [comment]
> then we need to notify everyone before merge this

I've already notified people on Discord, so they know that they should check this PR and test it together with their modules in use. As usual, some people will be helping us (and themselves) and do it before this gets merged, while others will not pay attention and complain when their modules will be broken but nothing we can do about it.

**Comment by UltraNix** [comment]
> > @Yehonal I made PRs in every modul that uses guids or is changes by that PR. Give me an example of modul that I forgot about?
> > And no - there is no possibility to offer "legacy" code or support, because of confilicts between modules and core. Sink or swim.
> 
> There are modules that are not published in our organization and I'm not sure about eluna as well. Have you checked that module? do the .lua files need to be updated?
> 
> If it's not possible to keep those legacy methods/variables then we need to notify everyone before merge this

I can do update only modules that are here, in this organization. There''s nothing I can do with code that I can't see.

**Comment by FrancescoBorzi** [comment]
> > > @Yehonal I made PRs in every modul that uses guids or is changes by that PR. Give me an example of modul that I forgot about?
> > > And no - there is no possibility to offer "legacy" code or support, because of confilicts between modules and core. Sink or swim.
> > 
> > 
> > There are modules that are not published in our organization and I'm not sure about eluna as well. Have you checked that module? do the .lua files need to be updated?
> > If it's not possible to keep those legacy methods/variables then we need to notify everyone before merge this
> 
> I can do update only modules that are here, in this organization. There''s nothing I can do with code that I can't see.
> 
> I will check Elune module.

@UltraNix it would be nice if you can check this PR with these modules, we use them in ChromieCraft and they are completely open-source:

- [mod-cfbg](https://github.com/azerothcore/mod-cfbg)
- [mod-duel-reset](https://github.com/azerothcore/mod-duel-reset)
- [mod-desertion-warnings](https://github.com/azerothcore/mod-desertion-warnings)
- [mod-ah-bot](https://github.com/azerothcore/mod-ah-bot)
- [mod-eluna-lua-engine](https://github.com/azerothcore/mod-eluna-lua-engine)
- [mod-server-auto-shutdown](https://github.com/azerothcore/mod-server-auto-shutdown)
- [lua-LevelUpReward](https://github.com/55Honey/Acore_LevelUpReward)
- [lua-CarbonCopy](https://github.com/55Honey/Acore_CarbonCopy)

**Comment by UltraNix** [comment]
All but the Elune and not from azerothcore should be fine.

**Comment by UltraNix** [comment]
Eluna done here: https://github.com/ElunaLuaEngine/Eluna/pull/357

@FrancescoBorzi Please, check https://github.com/55Honey/Acore_LevelUpReward and https://github.com/55Honey/Acore_CarbonCopy with this PR.

**Comment by Winfidonarleyan** [comment]
- Unknown my nickname
image

**Comment by UltraNix** [comment]
@Winfidonarleyan Fixed.

**Comment by UltraNix** [comment]
@FrancescoBorzi https://github.com/55Honey/Acore_LevelUpReward and https://github.com/55Honey/Acore_CarbonCopy are okay.

**Comment by Winfidonarleyan** [comment]
Okay, work

Tested:
- Hunter pets (loading, combat, stable master, rename)
- Items (create with professions, save/load, buy/sell)
- Gossips (with playername)
- Battlegrounds (wsg)

**Comment by FrancescoBorzi** [comment]
thanks @UltraNix for the excellent work and thanks @Winfidonarleyan for testing this

**Comment by Kitzunu** [comment]
I'd love this PR to get merged to get less conflicts for future PRs and so this PR don't need to be updated all the time... :D

**Comment by FrancescoBorzi** [comment]
@UltraNix can you please fix the conflicts one more time? then I'll merge it

**Comment by UltraNix** [comment]
@FrancescoBorzi Done.

**Comment by xIvan0ff** [comment]
Now all custom modules are broken. Please accept the PRs for custom modules!

**Comment by FrancescoBorzi** [comment]
> Now all custom modules are broken. Please accept the PRs for custom modules!

I've merged almost all of them, a couple of them have failing build though so I haven't merged them yet

**Comment by hahalmtdhl** [comment]
Connected PRs: And CFBG moudle

**Comment by FrancescoBorzi** [comment]
@UltraNix can you please take care of fixing the CFBG module?

**Comment by hahalmtdhl** [comment]
It's going to be a very big submission, including custom modules. Luahook is currently being modified.

**Comment by hahalmtdhl** [comment]
> 这将是一个非常大的提交，包括自定义模块。Luahook目前正在修改中。

Basically, the getcounter method is used to get the integer value of an object

**Comment by UltraNix** [comment]
@FrancescoBorzi You mean https://github.com/azerothcore/mod-cfbg? It's already merged.

**Comment by FrancescoBorzi** [comment]
thanks @UltraNix 

@hahalmtdhl can you update to latest CFBG version and try again?

**Comment by LannyE** [comment]
Solocraft https://github.com/azerothcore/mod-solocraft is failing to load after this.  It also has a PR that is waiting to be merged for an issue.

**Comment by LannyE** [comment]
Nevermind, I think I figured it out what this changes.  Please review and/or merge the PR I have waiting for the Solocraft mod.

**Pull request merged**